Implement a new feature

scinart opened the pull request:
A macro ``LOGURU_WITH_FILEABS`` is added.

when ``LOGURU_WITH_FILEABS`` is defined, a check of file change will be performed on every call to ``file_log``. If the file is moved, or inode changes, file is reopened using the same ``FileMode`` as is done by ``add_file``.

``file_log`` is made sequential using pthread lock since every call of ``file_log`` may reopen the file and invalidate the file descriptor. Hence calling ``file_log`` from multiple threads is OK.

Such a scheme is useful if you have a daemon program that moves the log file every 24 hours and expects new file to be created.

If one thread calls remove_callback while others are logging into this file, an error will probably occur. But I will let it be since I guess the very problem exists in the original design.

emilk commented on loguru.hpp:
```diff
@@ -1281,6 +1281,19 @@ This will define all the Loguru functions so that the linker may find them.
+#ifdef LOGURU_WITH_FILEABS
+	struct FileAbs
+	{
+		pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
```
All logging is already protected by the global `s_mutex`, so this lock is redundant afaict.

scinart replied:
Thanks, I'll try to fix it.

emilk commented on loguru.hpp:
```diff
@@ -1379,29 +1392,100 @@ namespace loguru
+		pthread_mutex_lock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+		file_reopen(user_data);
+		FILE* file = to_file(user_data);
+		if(!file) {
```
Please follow the convention `if (` (note the space) here and elsewhere

scinart replied:
OK, I'll pay more attention to the indention.

emilk commented on loguru.hpp:
```diff
@@ -1379,29 +1392,100 @@ namespace loguru
+				stat(file_abs->path, &file_abs->st);
+			}
+			// adapted version, only add to callback vector s_callbacks;
+			add_callback_abs(file_abs->path, file_log, file_abs, file_abs->verbosity, file_close, file_flush);
```
Can we get rid of the `remove_callback_abs` and `add_callback_abs` somehow? They contribute a lot of code and complexity, and reorders the order of the logging callbacks, which can be surprising.

I'm assuming they are only there so you can log things during file reopening, without infinite recursion? If so, I feel something much more simpler would be to have a single `bool is_reopening = false;` in `FileAbs` which is set to `true` during the logic above. Thus we could check `is_reopening` in `file_reopen` to avoid infinite recursion.

scinart replied:
I'm in a rush to come up with this solution. My solution does seem naive. I'll fix it. Thanks.

emilk commented on loguru.hpp:
```diff
@@ -1379,29 +1392,100 @@ namespace loguru
+		if (!file_abs->fp || (ret = stat(file_abs->path, &st)) == -1 || (st.st_ino != file_abs->st.st_ino)) {
+			remove_callback_abs(file_abs->path); //adapted version, only remove from callback vector s_callbacks;
+			if (ret < 0) {
+				LOG_F(INFO, "Reopening file '%s' due to %m", file_abs->path); // still logging to previous fd;
```
This is the first time I see `%m` - is that supported by Clang? I'd prefer more portable code - in particular, this is cross-platform (and thread-safe):

```
const auto why = errno_as_text();
LOG_F(INFO, "Reopening file '%s' due to %s", file_abs->path, why.c_str());
```

emilk commented on loguru.hpp:
```diff
@@ -1379,29 +1392,100 @@ namespace loguru
+			}
+			file_abs->fp = fopen(file_abs->path, file_abs->mode_str);
+			if (!file_abs->fp) {
+				LOG_F(ERROR, "Failed to open '%s'", file_abs->path);
```
Will this lead to retrying every time we log? Is this desired?

scinart replied:
I intended to retry everytime.

emilk commented on loguru.hpp:
```diff
@@ -71,7 +71,7 @@ Website: www.ilikebigbits.com
 
 	// Or just go with what Loguru suggests:
-	char log_path[1024];
+	char log_path[PATH_MAX];
```
:+1: here and elsewhere

emilk commented on loguru.hpp:
```diff
@@ -1759,11 +1842,21 @@ namespace loguru
 		const char* mode_str = (mode == FileMode::Truncate ? "w" : "a");
 		auto file = fopen(path, mode_str);
+#ifdef LOGURU_WITH_FILEABS
+		FileAbs* file_abs = new FileAbs(); // this is deleted file_close;
```
nit: `this is deleted in file_close`

scinart replied:
Uh, ``file_log`` etc uses ``void* user_data`` as a parameter. I don't know how to do better.

emilk replied:
Sorry for being unclear - I only meant the comment was missing the word "in" =)

emilk commented:
This looks really promising - thanks for your contribution! Please address my comments, and I'll merge this.

scinart commented:
Hi, I've updated my branch.

I guess trying to reopen a file every time we call file_log may not be a good idea, so I added ``const auto check_interval`` and call file_reopen every ``check_interval``.

It is currently set to 0 seconds in the code. But adding another macro for this seems to lead to more perplexity. Please decide how to handle this.

emilk commented:
LGTM - thank's for your contribution!
